apps/src/applab/firebaseCounters.js

```diff
@@ -69,6 +69,45 @@ export function updateTableCounters(tableName, rowCountChange, updateNextId) {
 }
 
 /**
+ * Increment or reset the rate-limiting counters associated with the interval.
+ * @param {number} maxWriteCount
+ * @param {string} interval
+ * @param {number} currentTimeMs
+ * @returns {Promise}
+ */
+function incrementIntervalCounters(maxWriteCount, interval, currentTimeMs) {
+  const limitRef = getDatabase(Applab.channelId).child(`counters/limits/${interval}`);
+  return limitRef.transaction(limitData => {
+    limitData = limitData || {};
+    limitData.lastResetTime = limitData.lastResetTime || 0;
+    limitData.writeCount = (limitData.writeCount || 0) + 1;
+    if (limitData.writeCount <= maxWriteCount) {
+      return limitData;
+    } else if (limitData.lastResetTime + interval * 1000 < currentTimeMs) {
+      // The maximum number of writes has been exceeded in more than `interval` seconds.
+      // Reset the counters.
+      limitData.writeCount = 1;
+      limitData.lastResetTime = Firebase.ServerValue.TIMESTAMP;
+      return limitData;
+    } else {
+      // The maximum number of writes has been exceeded in less than `interval` seconds.
+      // Abort the transaction.
+      return;
+    }
+  }).then(transactionData => {
+    if (!transactionData.committed) {
```

Comment:
Nope - the behavior you're describing sounds reasonable, but what actually happens when the `updateFunction` returns `undefined` is that the success handler is called with `committed` equal to `false`: https://www.firebase.com/docs/web/api/firebase/transaction.html

> onComplete Function (optional)
A callback function that will be called when the transaction completes. The callback is passed three arguments: a possibly-null error, a boolean indicating whether the transaction was committed, and a DataSnapshot indicating the final result. If the transaction failed abnormally, the first argument will be an Error object indicating the failure cause. **If the transaction finished normally, but no data was committed because no data was returned from updateFunction, then second argument (committed) will be false.** If the transaction completed and committed data to Firebase, committed will be true.

Promises were a bit of an afterthought in firebase and are documented in this blog post: https://firebase.googleblog.com/2016/01/keeping-our-promises-and-callbacks_76.html to summarize:

> the result of the transaction method is a Promise for an object with two fields: `committed` and `snapshot`. These map to the two parameters passed to the `onComplete` callback.